src/object/srv_obj.c

```diff
@@ -791,58 +734,235 @@ ds_obj_check_dtx_config(uint32_t flags, struct dtx_id *dti)
+		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
+			rpc_map_ver, map_ver);
+		if (obj_modify_rpc(opc))
+			D_GOTO(out_put, rc = -DER_STALE);
+		/* It is harmless if fetch with old pool map version. */
+	} else if (rpc_map_ver > map_ver) {
+		/* XXX: Client (or leader replica) has newer pool map than
+		 *	current replica. Two possibile cases:
+		 *
+		 *	1. The current replica was the old leader if with
+		 *	   the old pool map version. According to current
+		 *	   leader election algorithm, it is still the new
+		 *	   leader with the new pool map version. Since no
+		 *	   leader switch, the unmatched pool version will
+		 *	   not affect DTX related availability check.
+		 *
+		 *	2. The current replica was NOT the old leader if
+		 *	   with the old pool map version. But it becomes
+		 *	   the new leader with the new pool map version.
+		 *	   In the subsequent modificaiton, it may hit
+		 *	   some 'prepared' DTX when make availability
+		 *	   check, it will return -DER_INPROGRESS that
+		 *	   will cause client to retry. It is possible
+		 *	   that the pool map version event arrives at
+		 *	   this server during the client retry. It is
+		 *	   inefficient, but harmless.
+		 */
+	}
+
+	/* Check DTX configuration. */
+	if ((flags != 0 || dti != NULL) && obj_modify_rpc(opc) &&
+	     !daos_oc_echo_type(daos_obj_id2class(oid.id_pub))) {
+		rc = ds_obj_check_dtx_config(flags, dti);
+		if (rc != 0)
+			D_GOTO(out_put, rc);
+	}
+
+out_put:
+	if (rc) {
+		if (*contp != NULL) {
+			ds_cont_child_put(*contp);
+			*contp = NULL;
+		}
+		if (*hdlp != NULL) {
+			ds_cont_hdl_put(*hdlp);
+			*hdlp = NULL;
+		}
+	}
+
+	return rc;
+}
+
+void
+ds_obj_tgt_update_handler(crt_rpc_t *rpc)
+{
+	struct obj_rw_in		*orw = crt_req_get(rpc);
+	struct obj_rw_out		*orwo = crt_reply_get(rpc);
+	struct ds_cont_hdl		*cont_hdl = NULL;
+	struct ds_cont_child		*cont = NULL;
+	struct dtx_handle		*dth = NULL;
+	struct dtx_conflict_entry	 conflict = { 0 };
+	uint32_t			 map_ver = 0;
+	int				 rc;
+
+	D_ASSERT(orw != NULL);
+	D_ASSERT(orwo != NULL);
+
+	rc = ds_pre_check(orw->orw_oid, orw->orw_map_ver, orw->orw_co_hdl,
+			  orw->orw_co_uuid, opc_get(rpc->cr_opc),
+			  orw->orw_flags, &orw->orw_dti,
+			  &cont_hdl, &cont);
+	if (rc)
+		goto out;
+
+	D_ASSERT(cont_hdl->sch_pool != NULL);
+	map_ver = cont_hdl->sch_pool->spc_map_version;
+
+	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
+		DF_U64", pool ver %u/%u with "DF_DTI".\n",
+		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
+		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
+		dss_get_module_info()->dmi_tgt_id,
+		dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
+
+	/* Handle resend. */
+	if (orw->orw_flags & ORF_RESEND) {
+		rc = dtx_handle_resend(cont->sc_hdl, &orw->orw_oid,
+				       &orw->orw_dti,
+				       orw->orw_dkey_hash, false);
+		if (rc == -DER_ALREADY)
+			D_GOTO(out, rc = 0);
```

Comment:
If (rc == 0), then it is resend case, we should skip local update under such case.

Reply:
Hmm, I just followed the current master implementation here. Actually I am not sure I understand the current implementation here.
So rc == 0 means the DTX is init and prepared? == -DALREADY means the dtx has been committed. so we all should skip the local update in these cases? i.e. they are the same for resend, then could we all return 0 here. or I miss sth?

Reply:
" == 0" means resent, but the DTX has not been committed yet. For non-leader, it does not need to re-do the local modification.

" == -DALREADY" means resend, the DTX has been committed already, not need to re-do the local modification.

" == - DER_NONEXIST" means not resent, needs to to the local modification.

others are failure cases.